examples/gno.land/p/demo/microblog/microblog.gno

```diff
@@ -0,0 +1,174 @@
+package microblog
+
+import (
+	"errors"
+	"std"
+	"strings"
+	"time"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
+)
+
+type Microblog struct {
+	Title  string
+	Prefix string   // i.e. r/gnoland/blog:
+	Pages  avl.Tree // author (string) -> Page
+}
+
+func NewMicroblog(title string, prefix string) (m *Microblog) {
+	return &Microblog{
+		Title:  title,
+		Prefix: prefix,
+		Pages:  avl.Tree{},
+	}
+}
+
+func (m *Microblog) GetPages() []*Page {
+	var (
+		pages = make([]*Page, m.Pages.Size())
+		index = 0
+	)
+
+	m.Pages.Iterate("", "", func(page *avl.Node) bool {
+		pages[index] = page.Value().(*Page)
+		index++
+		return false
+	})
+
+	// TODO: sort by last updated
+
+	return pages
+}
+
+func (m *Microblog) Render(path string) string {
+	parts := strings.Split(path, "/")
+
+	isHome := path == ""
+	isUser := len(parts) == 1
+
+	switch {
+	case isHome:
+		output := ufmt.Sprintf("# %s\n\n", m.Title)
+		output += "# pages\n\n"
+
+		for _, page := range m.GetPages() {
+			output += ufmt.Sprintf("- [%s](%s%s)\n", page.Author.String(), m.Prefix, page.Author.String())
+		}
+
+		return output
+
+	case isUser:
+		user := parts[0]
+
+		silo, found := m.Pages.Get(user)
+		if !found {
+			return "404"
+		}
+
+		return (silo.(*Page)).String()
+	}
+
+	return "404"
+}
+
+func (m *Microblog) NewPage() error {
+	author := std.GetOrigCaller()
+	_, found := m.Pages.Get(author.String())
+	if found {
+		return errors.New("author already exists")
+	}
+
+	m.Pages.Set(author.String(), &Page{
+		Author:    author,
+		CreatedAt: time.Now(),
+	})
+	return nil
+}
+
+func (m *Microblog) NewPost(text string) error {
+	author := std.GetOrigCaller()
+	_, found := m.Pages.Get(author.String())
+	if !found {
+		// make a new page for the new author
+		if err := m.NewPage(); err != nil {
+			return err
+		}
+	}
+
+	return m.GetPage(author.String()).NewPost(text)
+}
+
+func (m *Microblog) GetPage(author string) *Page {
+	silo, found := m.Pages.Get(author)
+	if !found {
+		return nil
```

Comment:
Some methods error out (like `NewPage`), but some like this one don't.
Which errors do you consider to be hard errors (that warrant and error), and which should be a noop?

Reply:
maybe using something like `var ErrNotFound error = errors.New("not found")`

Reply:
Thanks, I would rather always have errors, I got a little confused about when to use them in Gno. For this I broke it out the errors: https://github.com/schollz/gno/commit/ecf5fb64b88172ad9cd213256086a5a8587dc493